tests/integration/test_delayed_replica_failover/test.py

```diff
@@ -112,3 +113,39 @@ def test(started_cluster):
     load_balancing='in_order',
     max_replica_delay_for_distributed_queries=1
 ''').strip() == '2'
+
+
+# in some scenarios one of the replicas can misbehave in a very weird way
+# just like if the process listening the socket was stopped
+# that can happen in clouds due to some operations on instances / VMs
+# or under certain network conditions
+@pytest.mark.timeout(120)
+def test_stuck_replica(started_cluster):
+    node_1_2.query("INSERT INTO replicated VALUES ('2020-03-27', 1)")
+
+    node_2_1.query("INSERT INTO replicated VALUES ('2020-03-27', 2)")
+    node_2_2.query("INSERT INTO replicated VALUES ('2020-03-27', 3)")
+    node_2_2.query("SYSTEM SYNC REPLICA replicated")
+    node_2_1.query("SYSTEM SYNC REPLICA replicated")
+
+    cluster.pause_container('node_2_1')
+
+    result = instance_with_dist_table.query("SELECT hostName(), x FROM distributed ORDER BY hostName(), x SETTINGS load_balancing='in_order', prefer_localhost_replica=0")
+    assert TSV(result) == TSV('node_1_1\t1\nnode_2_2\t2\nnode_2_2\t3')
+    # сейчас: висит 5 минут (receive_timeout), остановить запрос с помощью ctrl+c или по max_execution_time - невозможно
+
+    # ожидаемое поведение:
+    #  убеждаемся что реплика не даёт никакого acknowkegment (можно какой-то доп райунтрип "ты жива? да",
+    #  или ack после отправки запроса по типу "запроос принят, буду обрабатывать") в течение
+    #  connect_timeout_with_failover_ms, и если ack нет - идём в другую (здоровую) реплику
```

Comment:
This won't help, because if the server can be suddenly paused before hello packet, it can also be paused after it, with the same result.

Reply:
heartbeats

Reply:
Adding more kinds of packets to the protocol doesn't help -- they are all subject to the same problem, heartbeat packets included. The problem is not in the protocol but in the fact that we work with replicas synchronously. Anyway, we have some understanding of what can be done about this case, AFAIR @tavplubix wanted to make a prototype but haven't got to it yet.

Reply:
Yeah, it should be async.